### PR TITLE
Add monitor for local storage PV

### PR DESCRIPTION
**Pull request by NickrenREN**
first commit: add monitor for local storage PVs

xref : [pv health monitoring proposal](https://docs.google.com/document/d/1WG51DZZeXyP50EKyhECYg5m5KEB4F0AhYBhA555_Gs4/edit)

TODOs:
- [x] add timestamp for PV marking
- [x] add label/selector for monitor to filter PVs
- [x] unmark PV if it recovers
- [ ] ensure that PV usage is not greater than PV capacity ?
- [ ]  provide different monitor methods for different devices ?

### CODE REVIEW

File: local-volume/provisioner/pkg/monitor/monitor.go [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+				monitor.checkStatus(pv)
+			}
+		}
+
+		time.Sleep(DefaultMonitorResyncPeriod)
+	}
+}
+
+// checkStatus checks pv health condition
+func (monitor *Monitor) checkStatus(pv *v1.PersistentVolume) {
```
Comment (NickrenREN) [inline local-volume/provisioner/pkg/monitor/monitor.go:208]: This needs to be discussed more

File: local-volume/provisioner/pkg/monitor/monitor.go [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+func (monitor *Monitor) HasRun() bool {
+	monitor.hasRunLock.Lock()
+	defer monitor.hasRunLock.Unlock()
+	return monitor.hasRun
+}
+
+func (monitor *Monitor) MonitorLocalVolumes() {
+	for {
+		if monitor.HasRun() {
+			for _, pv := range monitor.volumeMap {
```
Comment (walktall) [inline local-volume/provisioner/pkg/monitor/monitor.go]: `volumeMap` was protected by a mutex?

Comment (NickrenREN) [reply local-volume/provisioner/pkg/monitor/monitor.go]: yeah, and i'm thinking whether we can reuse the volumecache,  any problem ?

Comment (walktall) [reply local-volume/provisioner/pkg/monitor/monitor.go]: I think there should be a mutex check while looping this map.

By `volumecache `, you mean `Store` returned by `Informer`?

Comment (NickrenREN) [reply local-volume/provisioner/pkg/monitor/monitor.go]: No, not `store`,  volume cache in `RuntimeConfig `
And this is a `get` opeation, will not change the map,  maybe do not need to add mutex

Comment (walktall) [reply local-volume/provisioner/pkg/monitor/monitor.go]: Maps are not safe for concurrent use, no matter read or write operations.

Ref: https://blog.golang.org/go-maps-in-action Check the `Concurrency ` Part.

Comment (walktall) [reply local-volume/provisioner/pkg/monitor/monitor.go]: And, why don't we use the `Store.List` directly?

Comment (NickrenREN) [reply local-volume/provisioner/pkg/monitor/monitor.go]: `Store` will store all PVs there and I'm thinking just focus on PVs(local storage) belonging to that node.

Comment (NickrenREN) [reply local-volume/provisioner/pkg/monitor/monitor.go]: btw: PVs may be created manually. We also need to take this into consideration too.

File: local-volume/provisioner/pkg/monitor/monitor.go [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+	if len(mountPath) == 0 {
+		// can not find mount path, this may because: admin modify config(hostpath)
+		// mark PV and send a event
+		err = monitor.markPV(pv, HostPathNotExist, "yes")
+		if err != nil {
+			glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+		}
+		monitor.Recorder.Event(pv, v1.EventTypeWarning, HostPathNotExist, "host path config may be modified")
+		return
+	} else {
```
Comment (ScorpioCPH) [inline local-volume/provisioner/pkg/monitor/monitor.go]: nit: maybe this `else` is needless, as we have returned in the previous `if` statement.

File: local-volume/provisioner/pkg/monitor/monitor.go [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+
+	monitor.volumeLW = &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			return config.Client.CoreV1().PersistentVolumes().List(options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			return config.Client.CoreV1().PersistentVolumes().Watch(options)
+		},
+	}
+	_, monitor.volumeController = cache.NewInformer(
```
Comment (ScorpioCPH) [inline local-volume/provisioner/pkg/monitor/monitor.go]: nit: as suggested [here](https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md#guidelines), maybe we can use `NewSharedInformerFactory`.

Comment (ScorpioCPH) [reply local-volume/provisioner/pkg/monitor/monitor.go]: and how about rename `volumeController` -> `volumeInformer`

File: local-volume/provisioner/pkg/monitor/monitor.go [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/api/v1/helper"
+)
+
+const (
+	// DefaultResyncPeriod is used when option function ResyncPeriod is omitted
+	DefaultResyncPeriod = 15 * time.Second
+
+	DefaultMonitorResyncPeriod = 3 * time.Minute
```
Comment (ScorpioCPH) [inline local-volume/provisioner/pkg/monitor/monitor.go]: there are two ResyncPeriod, maybe we can rename `DefaultResyncPeriod` -> `DefaultInformerResyncPeriod`?

File: local-volume/provisioner/pkg/monitor/monitor.go [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+const (
+	HostPathNotExist  = "HostPathNotExist"
+	MisMatchedVolSize = "MisMatchedVolSize"
+)
+
+// Monitor checks PVs' health condition and taint them if they are unhealthy
+type Monitor struct {
+	*common.RuntimeConfig
+
+	resyncPeriod time.Duration
```
Comment (ScorpioCPH) [inline local-volume/provisioner/pkg/monitor/monitor.go]: informerResyncPeriod?

File: local-volume/provisioner/pkg/monitor/monitor.go [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+				return false, fmt.Errorf("NodeSelectorTerm %+v does not match node labels", term.MatchExpressions)
+			}
+		}
+	}
+	return true, nil
+}
+
+// markPV marks PV by adding annotation
+func (monitor *Monitor) markPV(pv *v1.PersistentVolume, ann, value string) error {
+
```
Comment (ScorpioCPH) [inline local-volume/provisioner/pkg/monitor/monitor.go]: nit: empty line

File: local-volume/provisioner/pkg/monitor/monitor.go [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+	if ok {
+		glog.V(4).Infof("PV: %s is ready marked using ann: %s", volumeClone.Name, ann)
+		return nil
+	}
+	metav1.SetMetaDataAnnotation(&volumeClone.ObjectMeta, ann, value)
+
+	var err error
+	var newVol *v1.PersistentVolume
+	// Try to update the PV object several times
+	for i := 0; i < UpdatePVRetryCount; i++ {
```
Comment (ScorpioCPH) [inline local-volume/provisioner/pkg/monitor/monitor.go:447]: maybe we can use a flag to check whether this updating is success?

File: local-volume/provisioner/pkg/monitor/monitor.go [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+			glog.V(4).Infof("updating PersistentVolume[%s] failed: %v", volumeClone.Name, err)
+			continue
+		}
+		monitor.localVolumeMap.UpdateLocalVolume(newVol)
+		glog.V(4).Infof("updating PersistentVolume[%s] successfully", newVol.Name)
+		return nil
+
+		time.Sleep(UpdatePVInterval)
+	}
+
```
Comment (ScorpioCPH) [inline local-volume/provisioner/pkg/monitor/monitor.go]: check the updating flag here and report an error if failed.

File: local-volume/provisioner/pkg/util/util.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+limitations under the License.
+*/
+
+package util
+
+import esUtil "github.com/kubernetes-incubator/external-storage/lib/util"
+
+// Round down the capacity to an easy to read value.
+func RoundDownCapacityPretty(capacityBytes int64) int64 {
+
```
Comment (ScorpioCPH) [inline local-volume/provisioner/pkg/util/util.go]: nit: empty line

File: local-volume/provisioner/pkg/monitor/monitor.go [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+		}
+	}
+	if len(mountPath) == 0 {
+		// can not find mount path, this may because: admin modify config(hostpath)
+		// mark PV and send a event
+		err = monitor.markPV(pv, HostPathNotExist, "yes")
+		if err != nil {
+			glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+		}
+		monitor.Recorder.Event(pv, v1.EventTypeWarning, HostPathNotExist, "host path config may be modified")
```
Comment (ScorpioCPH) [inline local-volume/provisioner/pkg/monitor/monitor.go]: where is `Recorder` defined?

**Review by ScorpioCPH — COMMENTED** [review]
generally lgtm.

File: local-volume/provisioner/pkg/monitor/monitor.go (hunk 11 above)
Comment (NickrenREN) [reply local-volume/provisioner/pkg/monitor/monitor.go]: at `StartLocalController` in provisioner/controller package

File: local-volume/provisioner/pkg/monitor/monitor.go (hunk 8 above)
Comment (NickrenREN) [reply local-volume/provisioner/pkg/monitor/monitor.go:447]: will add event

Comment (NickrenREN) [reply local-volume/provisioner/pkg/monitor/monitor.go:447]: added

File: local-volume/provisioner/pkg/util/util.go (hunk 10 above)
Comment (NickrenREN) [reply local-volume/provisioner/pkg/util/util.go]: Removed

File: local-volume/provisioner/pkg/monitor/monitor.go (hunk 9 above)
Comment (NickrenREN) [reply local-volume/provisioner/pkg/monitor/monitor.go]: ditto

File: local-volume/provisioner/pkg/monitor/monitor.go (hunk 7 above)
Comment (NickrenREN) [reply local-volume/provisioner/pkg/monitor/monitor.go]: done

File: local-volume/provisioner/pkg/monitor/monitor.go (hunk 4 above)
Comment (NickrenREN) [reply local-volume/provisioner/pkg/monitor/monitor.go]: The related package has not been imported yet for external-storage repo

File: local-volume/provisioner/pkg/monitor/monitor.go (hunk 5 above)
Comment (NickrenREN) [reply local-volume/provisioner/pkg/monitor/monitor.go]: done

**Pull request closed without merging**